pkg/driver/aws-ebs/aws_ebs_tags_controller.go

```diff
@@ -0,0 +1,483 @@
+		awsInfra := infra.Status.PlatformStatus.AWS
+		err := c.fetchPVsAndUpdateTags(awsInfra.ResourceTags, ec2Client)
+		if err != nil {
+			klog.Errorf("Error processing PVs for infrastructure update: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// fetchPVsAndUpdateTags retrieves all PVs and updates the AWS EBS tags in batches of 100
+func (c *EBSVolumeTagsController) fetchPVsAndUpdateTags(resourceTags []configv1.AWSResourceTag, ec2Client *ec2.EC2) error {
+	pvs, err := c.commonClient.KubeClient.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error fetching PVs: %v", err)
+	}
+
+	// Collect the volume IDs
+	var volumeIDs []string
+	for _, pv := range pvs.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == driverName {
+			volumeID := pv.Spec.CSI.VolumeHandle
+			volumeIDs = append(volumeIDs, volumeID)
+		}
+	}
+
+	// Process the volume IDs in batches of 100
+	const batchSize = 100
+	for i := 0; i < len(volumeIDs); i += batchSize {
+		end := i + batchSize
+		if end > len(volumeIDs) {
+			end = len(volumeIDs)
+		}
+		batch := volumeIDs[i:end]
+
+		err = c.updateEBSTags(batch, ec2Client, resourceTags)
+		if err != nil {
+			errorMessage := fmt.Sprintf("Error creating tags for volume %s: %v", volumeIDs, err)
+			c.handleFailure(volumeIDs, errorMessage)
+		}
+		c.handleSuccess(volumeIDs)
+	}
+
+	return nil
+}
+
+// updateEBSTags updates the tags of an AWS EBS volume
+func (c *EBSVolumeTagsController) updateEBSTags(volumeIDs []string, ec2Client *ec2.EC2, resourceTags []configv1.AWSResourceTag) error {
+
+	// Merge the existing tags with new resource tags
+	mergedTags := newAndUpdatedTags(resourceTags)
+
+	klog.Infof("Updating EBS tags for volume IDs %s with tags: %v", volumeIDs, mergedTags)
+
+	// Create or update the tags
+	_, err := ec2Client.CreateTags(&ec2.CreateTagsInput{
+		Resources: volumesIDsToResourceIDs(volumeIDs),
+		Tags:      mergedTags,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error creating tags for volume %s: %v", volumeIDs, err)
+	}
+
+	klog.Infof("Successfully updated tags for volumes %s", volumeIDs)
+	return nil
+}
+
+// newAndUpdatedTags adds and update existing AWS tags with new resource tags from OpenShift infrastructure
+func newAndUpdatedTags(resourceTags []configv1.AWSResourceTag) []*ec2.Tag {
+	// Convert map back to slice of ec2.Tag
+	var tags []*ec2.Tag
+	for _, tag := range resourceTags {
+		tags = append(tags, &ec2.Tag{
+			Key:   aws.String(tag.Key),
+			Value: aws.String(tag.Value),
+		})
+	}
+	return tags
+}
+
+func volumesIDsToResourceIDs(volumeIDs []string) []*string {
+	var resourceIDs []*string
+	for _, volumeID := range volumeIDs {
+		resourceIDs = append(resourceIDs, aws.String(volumeID))
+	}
+	return resourceIDs
+}
+
+// startFailedQueueWorker runs a worker that processes failed batches independently
+func (c *EBSVolumeTagsController) startFailedQueueWorker(ctx context.Context) {
```

Comment:
I am not sure if we need this function at all.

Reply:
So the reason I brought this change is I would like to retry the the batches which have failed to update tags!! Here I would like to update tags in a serial order(one By one) [Discussuion link](https://redhat-internal.slack.com/archives/C057L1GKYSK/p1729528491363079) for the volumes. I cannot use the similar sync function here or else you could suggest a better way for this!!

Reply:
If we are going to per-pv hash, then we are not going to need this right?

Reply:
The points I want to make sure here are 

1. I definitely need to batch volumes in order to not to hit throttling condition.
2. Now on failure we would need to retry!!
3. Since we are using batch APIs, so AWS SDK's APIs will give us error if any one of the VolumeID in the batch hits the any error(May be validation, Auth, Permission etc), All the VolumeIDs in that batch will not be able to update the tags! In this condition I would like to add a worker queue that will handle the the serial update of PV's tags and will retry to update the tags in exponential back-off time-period.

Since We need to figure out the trade-off between the either in first place we should update all PVs using AWS API's in serial order and retry using similar sync function or should we use Batch Volumes to be called in the sync Functions and later retry should be processed with the queue function serially until the queue is empty!!

Reply:
>I definitely need to batch volumes in order to not to hit throttling condition.

That is fine. You already batching PVs in `fetchPVsAndUpdateTags`.

>Now on failure we would need to retry!!

We are unnecessarily complicating this. The controller is going to `resync` every 20 minutes anyways, so it will try to tag all the PVs which aren't tagged. So do we even need to keep separate queue for failed PVs? I am also afraid that, your failed worker queue is going to race with normal controller resync.

Reply:
Hey @gnufied I completely agree that this retrying the failure is unnecessarily making the changes complicated.

@jsafrane I had a chat with @TrilokGeer regarding this, IMHO we should drop the idea to add degraded condition on tags update failure. As the cluster should not be degraded for failure to apply Tags. as the sync will anyhow retry to tag the volumes withing resync period. 
Also anyways we are emitting the warning events from the this controller if the tags update is failed. so user will know that the tags update has been failed due the the certain reason.

/cc @TrilokGeer 
Can you also put your views on this!!

Reply:
BTW are we planning to backport this PR to older releases?

Reply:
@gnufied I guess we would need to backport this to 4.17. [Slack Thread](https://redhat-internal.slack.com/archives/C057L1GKYSK/p1727275576645219?thread_ts=1727193184.791609&cid=C057L1GKYSK).

Reply:
So I noticed that you removed failed worker logic. The thing is - I was talking to @jsafrane offline and he has me convinced that, we do need some kind of additional logic so as we can try retagging of "failed" PVs one-by-one, rather than in a batch. This will ensure that, one bad apple in a batch doesn't prevent tagging of rest of the PVs.

But - we need to be careful when doing this. 
- We should make sure that, PVs which will be tagged via failed worker, doesn't get processed via regular controller resync (so no race). 
- I would move the entire failed worker code in a separate file.

Reply:
Ack; 
I believe we can remove the  `ResyncEvery` parameter from the controller builder as this is a overkill for us why do we want to run the sync function in every 20 mins if nothing has changed in the resource Tags.

Alternatively, If you really think resync is important here then, I think that we can add the handle the race condition by using another annotation in PVs for Tagging status and filtering based on tagshash and status in the annotation. But this will also cost us some volume Update calls and further need to handle cases where we are not able to update the status within batchs.